flang/lib/Lower/CallInterface.cpp

```diff
@@ -461,13 +461,37 @@ class Fortran::lower::CallInterfaceImpl {
     }
   }
 
+  // Some dummy argument attributes prevent it to be used with implicit
+  // interfaces but that odes not imply that a box is needed for those. For
+  // instance, scalar intrinsic that are OPTIONAL require an explicit interface,
+  // but they can still be passed by reference.
+  bool dummyRequiresBox(
+      const Fortran::evaluate::characteristics::DummyDataObject &obj) {
+    using ShapeAttr = Fortran::evaluate::characteristics::TypeAndShape::Attr;
+    using ShapeAttrs = Fortran::evaluate::characteristics::TypeAndShape::Attrs;
+    constexpr ShapeAttrs shapeRequiringBox = {
+        ShapeAttr::AssumedShape, ShapeAttr::DeferredShape,
+        ShapeAttr::AssumedRank, ShapeAttr::Coarray};
```

Comment:
Are coarrays always going to require a box?

Reply:
Most likely. I did not dwell to much into coarrays, but I think that opencoarrays library is using descriptors for instance ([see `gfc_descriptor_t` here](https://github.com/sourceryinstitute/OpenCoarrays/blob/main/src/libcaf.h#L258-L300)).